cdc/sink/codec/avro.go

```diff
@@ -238,233 +400,260 @@ func ColumnInfoToAvroSchema(name string, columnInfo []*model.Column) (string, er
-	Precision:   64, // enough to hold all values and is the default precision of confluent schema registry
-	Scale:       0,
+	log.Debug("rowToAvroData", zap.Reflect("data", ret))
+	return ret, nil
 }
 
-func getAvroDataTypeFromColumn(col *model.Column) (interface{}, error) {
-	log.Info("DEBUG: getAvroDataTypeFromColumn", zap.Reflect("col", col))
+func columnToAvroSchema(
+	col *model.Column,
+	ft *types.FieldType,
+	decimalHandlingMode string,
+	bigintUnsignedHandlingMode string,
+) (interface{}, error) {
+	tt := getTiDBTypeFromColumn(col)
 	switch col.Type {
-	case mysql.TypeFloat:
-		return "float", nil
-	case mysql.TypeDouble:
-		return "double", nil
-	case mysql.TypeVarchar, mysql.TypeString, mysql.TypeVarString:
-		if col.Flag.IsBinary() {
-			return "bytes", nil
+	case mysql.TypeTiny, mysql.TypeShort, mysql.TypeInt24: // BOOL/TINYINT/SMALLINT/MEDIUMINT
+		return avroSchema{
+			Type:       "int",
+			Parameters: map[string]string{tidbType: tt},
+		}, nil
+	case mysql.TypeLong: // INT
+		if col.Flag.IsUnsigned() {
+			return avroSchema{
+				Type:       "long",
+				Parameters: map[string]string{tidbType: tt},
+			}, nil
 		}
-		return "string", nil
-	case mysql.TypeDate, mysql.TypeDatetime, mysql.TypeTimestamp:
-		return avroLogicalType{
-			Type:        "long",
-			LogicalType: timestampMillis,
+		return avroSchema{
+			Type:       "int",
+			Parameters: map[string]string{tidbType: tt},
 		}, nil
-	case mysql.TypeDuration:
-		return avroLogicalType{
-			Type:        "int",
-			LogicalType: timeMillis,
+	case mysql.TypeLonglong: // BIGINT
+		if col.Flag.IsUnsigned() && bigintUnsignedHandlingMode == "string" {
+			return avroSchema{
+				Type:       "string",
+				Parameters: map[string]string{tidbType: tt},
+			}, nil
+		}
+		return avroSchema{
+			Type:       "long",
+			Parameters: map[string]string{tidbType: tt},
+		}, nil
+	case mysql.TypeFloat, mysql.TypeDouble:
+		return avroSchema{
+			Type:       "double",
+			Parameters: map[string]string{tidbType: tt},
 		}, nil
-	case mysql.TypeEnum:
-		return "string", nil
-	case mysql.TypeSet:
-		return "string", nil
 	case mysql.TypeBit:
-		return unsignedLongAvroType, nil
-	case mysql.TypeNewDecimal:
-		return "string", nil
-	case mysql.TypeTiny, mysql.TypeShort, mysql.TypeInt24:
-		return "int", nil
-	case mysql.TypeLong:
-		if col.Flag.IsUnsigned() {
-			return "long", nil
+		displayFlen := ft.Flen
+		if displayFlen == -1 {
+			displayFlen, _ = mysql.GetDefaultFieldLengthAndDecimal(col.Type)
 		}
-		return "int", nil
-	case mysql.TypeLonglong:
-		if col.Flag.IsUnsigned() {
-			return unsignedLongAvroType, nil
+		return avroSchema{
+			Type: "bytes",
+			Parameters: map[string]string{
+				tidbType: tt,
+				"length": strconv.Itoa(displayFlen),
+			},
+		}, nil
+	case mysql.TypeNewDecimal:
+		if decimalHandlingMode == "precise" {
+			defaultFlen, defaultDecimal := mysql.GetDefaultFieldLengthAndDecimal(ft.Tp)
+			displayFlen, displayDecimal := ft.Flen, ft.Decimal
+			if displayFlen == -1 {
```

Comment:
`-1` confusing.

Reply:
:-( that's what TiDB returns, not much we could do. I'd like readers to learn the complex TiDB default handling logic.